### PR TITLE
Add Tensor class

### PR DESCRIPTION
**Pull request by apertovs**

### CODE REVIEW

**Comment by apertovs** [comment]
@l-bat, @TolyaTalamanov, @KSTconsult, @RodionovF Have a look, please.

File: include/tensor.hpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
+#include <vector>
+#include <memory>
+
+namespace deepworks {
+
+using Shape = std::vector<size_t>;
```
Comment (TolyaTalamanov) [inline include/tensor.hpp:6]: Shape can be `{-1, ...}`
I think it should be `class Shape : private std::vector<int>`

@l-bat

File: include/tensor.hpp [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+using Strides = std::vector<size_t>;
+
+class Tensor {
+public:
+    using Type = float;
+    explicit Tensor(const Shape &shape);
```
Comment (TolyaTalamanov) [inline include/tensor.hpp:12]: Consider to add `Shape&&` overload

File: include/tensor.hpp [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    Tensor();
```
Comment (TolyaTalamanov) [inline include/tensor.hpp:13]: Is it really necessary ? @l-bat

File: include/tensor.hpp [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+    Tensor(const Tensor&) = default;
```
Comment (TolyaTalamanov) [inline include/tensor.hpp:15]: `Tensor(const Tensor&)`
`Tensor(Tensor&&)` 

generated by compiler, there is no necessary to put `default`

File: include/tensor.hpp [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    Tensor(Tensor&&) = default;
+    Tensor& operator=(const Tensor&) = default;
```
Comment (TolyaTalamanov) [inline include/tensor.hpp:17]: The same

File: src/tensor.cpp [hunk 6]
```diff
@@ -0,0 +1,54 @@
+#include <tensor.hpp>
+#include <cstring>
```
Comment (TolyaTalamanov) [inline src/tensor.cpp:2]: `inlcude<string>` ?

File: src/tensor.cpp [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+#include <numeric>
+
+namespace deepworks {
+
+struct Tensor::Descriptor {
+    Shape _shape{};
```
Comment (TolyaTalamanov) [inline src/tensor.cpp:8]: `Shape _shape` would be enough

File: src/tensor.cpp [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    Strides _strides{};
+    Type *_data{nullptr};
+};
+
+Tensor::Tensor() : Tensor(Shape{1}) {}
```
Comment (TolyaTalamanov) [inline src/tensor.cpp:13]: From my point of view it is not obvious, suggest to remove such `ctor`

File: src/tensor.cpp [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+Tensor::Tensor(const Shape &shape) {
+    size_t n_elements = 1;
+    for (auto dim : shape) {
```
Comment (TolyaTalamanov) [inline src/tensor.cpp:17]: How to compute strides if shape have `-1` ?
@l-bat

File: src/tensor.cpp [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+    Strides strides(shape.size());
+
+    size_t initial_stride = 1;
+    auto dim_it = shape.rbegin();
+    for (auto stride_it = strides.rbegin(); stride_it != strides.rend(); ++stride_it, ++dim_it) {
+        *stride_it = initial_stride;
+        initial_stride *= *dim_it;
+    }
+    _descriptor = std::make_shared<Descriptor>(Descriptor{shape, strides, new Type[n_elements]});
```
Comment (TolyaTalamanov) [inline src/tensor.cpp:29]: You can use `reset` to avoid allocation + move in that case

File: src/tensor.cpp [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    tensor._descriptor->_shape = _descriptor->_shape;
+    tensor._descriptor->_strides = _descriptor->_strides;
+
+    size_t n_elements = std::accumulate(_descriptor->_shape.begin(),
+                                        _descriptor->_shape.end(),
+                                        1,
+                                        std::multiplies<>());
+    std::memcpy(tensor._descriptor->_data, _descriptor->_data, n_elements);
+}
+} // namespace deepworks
```
Comment (TolyaTalamanov) [inline src/tensor.cpp:54]: New line ?

File: src/tensor.cpp [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+const Strides &Tensor::strides() const {
+    return _descriptor->_strides;
+}
+
+const Shape &Tensor::shape() const {
+    return _descriptor->_shape;
+}
+
+void Tensor::copyTo(Tensor &tensor) {
+    tensor._descriptor->_shape = _descriptor->_shape;
```
Comment (TolyaTalamanov) [inline src/tensor.cpp:45]: Should we check that total numbers of elements is equal to passed tensor ?

File: src/tensor.cpp [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+    tensor._descriptor->_strides = _descriptor->_strides;
+
+    size_t n_elements = std::accumulate(_descriptor->_shape.begin(),
+                                        _descriptor->_shape.end(),
+                                        1,
+                                        std::multiplies<>());
+    std::memcpy(tensor._descriptor->_data, _descriptor->_data, n_elements);
```
Comment (TolyaTalamanov) [inline src/tensor.cpp:52]: `std::copy`

File: tests/unit/test_tensor.cpp [hunk 14]
```diff
@@ -0,0 +1,66 @@
+#include <deepworks.hpp>
+#include <gtest/gtest.h>
+#include <tensor.hpp>
+
+TEST(TensorTest, Strides) {
+    struct CaseData {
+        deepworks::Shape actualShape;
```
Comment (TolyaTalamanov) [inline tests/unit/test_tensor.cpp:7]: let's keep consistency and not use came case

File: tests/unit/test_tensor.cpp [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+    EXPECT_EQ(tensor.data(), srcTensor.data());
+}
+
+TEST(TensorTest, CopyTo) {
+    deepworks::Tensor srcTensor({1, 3, 224, 224});
+    for (size_t index = 0; index < 1 * 3 * 224 * 224; ++index) {
+        srcTensor.data()[index] = index;
+    }
+
+    deepworks::Tensor dstTensor;
```
Comment (TolyaTalamanov) [inline tests/unit/test_tensor.cpp:57]: Maybe for that case the `default` ctor is really needed, let's ask others...

**Review by TolyaTalamanov — COMMENTED** [review]
Overall looks great!!!
Only cosmetic changes are required.
Let's decide about `strides` and go merge it.

File: include/tensor.hpp [hunk 16]
```diff
@@ -0,0 +1,30 @@
+#include <vector>
```
Comment (l-bat) [inline include/tensor.hpp:1]: `#pragma once`  is missed

File: include/tensor.hpp (hunk 1 above)
Comment (l-bat) [reply include/tensor.hpp:6]: I agree with @TolyaTalamanov, we can do like  https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/include/ngraph/shape.hpp#L30.
Add  operator<< for Shape

**Review by l-bat — APPROVED** [review]

File: src/tensor.cpp [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+        throw std::runtime_error("Cannot copy tensor. Inconsistent number of elements");
+    }
+    tensor._descriptor->_shape = _descriptor->_shape;
+    tensor._descriptor->_strides = _descriptor->_strides;
+
+    std::copy_n(_descriptor->_data, n_elements, tensor._descriptor->_data);
+}
+
+Tensor::~Tensor() {
+    if (_descriptor.use_count() == 1) {
```
Comment (TolyaTalamanov) [inline src/tensor.cpp:79]: ~Description() implementation would be enough

File: include/tensor.hpp [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+    Tensor();
+    explicit Tensor(const Shape &shape);
+
+    void copyTo(Tensor &tensor);
+    Type *data();
+    size_t totalElements() const;
+    const Strides &strides() const;
+    const Shape &shape() const;
+
+    ~Tensor();
```
Comment (TolyaTalamanov) [inline include/tensor.hpp:24]: Unnecessary here, should be `Descriptor` responsibility

File: src/tensor.cpp [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+    m_data = new Type[total_elements];
+    m_created = true;
+}
+
+Tensor::Descriptor::~Descriptor() {
+    delete [] m_data;
+}
+
+/* Tensor */
+Tensor::Tensor() {
```
Comment (TolyaTalamanov) [inline src/tensor.cpp:64]: Now, you can do it like this: `Tensor::Tensor() : m_descriptor(new Descriptor()) {}`

File: src/tensor.cpp [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+    m_descriptor.reset(new Descriptor());
+}
+
+Tensor::Tensor(const Shape &shape) {
+    m_descriptor.reset(new Descriptor(shape));
+}
+
+void Tensor::copyTo(Tensor &tensor) {
```
Comment (TolyaTalamanov) [inline src/tensor.cpp:72]: What if

`t.copyTo(Tensor(shape, data))`

Let it be just `void Tensor::copyTo(Tensor tensor)`

File: include/tensor.hpp [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+    using Type = float;
+
+    Tensor();
+    explicit Tensor(const Shape &shape);
+
+    void copyTo(Tensor &tensor);
+    Type *data();
+    size_t total() const;
+    void create();
+    bool isCreated() const;
```
Comment (TolyaTalamanov) [inline include/tensor.hpp:22]: not get the logic for this, it's just `empty()`

File: src/tensor.cpp [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+
+    m_shape = descriptor.m_shape;
+    m_strides = descriptor.m_strides;
+
+    std::copy_n(m_data, m_total, descriptor.m_data);
+}
+
+void Tensor::Descriptor::allocate(const Shape &shape) {
+    if (m_data != nullptr) {
+        throw std::runtime_error("Tensor already allocated, cannot allocate twice.");
```
Comment (TolyaTalamanov) [inline src/tensor.cpp:31]: Just wondering why not ? But ok, let it be

**Comment by TolyaTalamanov** [comment]
@l-bat @RodionovF @KSTconsult Do you have any questions ?

File: src/tensor.cpp [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+    }
+    m_shape = shape;
+    m_total = std::accumulate(shape.begin(),
+                              shape.end(),
+                              1,
+                              std::multiplies<>());
+    m_data = new Type[m_total];
```
Comment (KSTconsult) [inline src/tensor.cpp:38]: What if shape is a vector = {-1}? Then m_total will be -1 also. You can`t allocate memory for negative size.

**Comment by RodionovF** [comment]
lgtm :+1:

**Comment by TolyaTalamanov** [comment]
I'd move `Descriptor` to the `*.cpp` to hide it from user, but let's do it later.

Thanks! :fire:

**Pull request merged**